File: BrainExtraction/Infant/T1T2/batchGenerator.R
batchGenerator <- function( batchSize = 32L,
                            imageSize = c( 64, 64, 64 ),
                            template = NULL,
                            images = NULL,
                            brainMasks = NULL,
                            doRandomContralateralFlips = TRUE,
                            doDataAugmentation = TRUE
 )
{                                     

  if( is.null( template ) )
    {
    stop( "No reference template specified." )
    }
  if( is.null( images ) )
    {
    stop( "Input images must be specified." )
    }
  if( is.null( brainMasks ) )
    {
    stop( "Input masks must be specified." )
    }

  currentPassCount <- 0L

  function()
    {
    # Shuffle the data after each complete pass

    if( ( currentPassCount + batchSize ) >= length( images ) )
      {
      # shuffle the source data
      sampleIndices <- sample( length( images ) )
      images <- images[sampleIndices]
      brainMasks <- brainMasks[sampleIndices]

      currentPassCount <- 0L
      }

    batchIndices <- currentPassCount + 1L:batchSize

    batchImages <- images[batchIndices]
    batchBrainMasks <- brainMasks[batchIndices]

    X <- array( data = 0, dim = c( batchSize, imageSize, 2 ) )    
    Y <- array( data = 0L, dim = c( batchSize, imageSize ) )

    currentPassCount <<- currentPassCount + batchSize

    # pb <- txtProgressBar( min = 0, max = batchSize, style = 3 )

    i <- 1
    while( i <= batchSize )
      {
      # setTxtProgressBar( pb, i )

      t1 <- NULL 
      t2 <- NULL 
      batchT1 <- antsImageRead( batchImages[[i]][1] )
      batchT2 <- antsImageRead( batchImages[[i]][2] )
      batchBrainMask <- thresholdImage( antsImageRead( batchBrainMasks[i] ), 0.5, 1.0, 1, 0 )

      if( doRandomContralateralFlips && sample( c( TRUE, FALSE ) ) )
        {
        t1 <- reflectImage( batchT1, axis = 0, verbose = FALSE )  
        t2 <- reflectImage( batchT2, axis = 0, verbose = FALSE )  
        brainMask <- reflectImage( batchBrainMask, axis = 0, verbose = FALSE )
        } else {
        t1 <- batchT1
        t2 <- batchT2
        brainMask <- batchBrainMask
        }

      centerOfMassTemplate <- getCenterOfMass( template )
      centerOfMassImage <- getCenterOfMass( brainMask )
      xfrm <- createAntsrTransform( type = "Euler3DTransform",
        center = centerOfMassTemplate,
        translation = centerOfMassImage - centerOfMassTemplate )
      warpedT1 <- applyAntsrTransformToImage( xfrm, t1, template )
      warpedT2 <- applyAntsrTransformToImage( xfrm, t2, template )
      warpedBrainMask <- thresholdImage( applyAntsrTransformToImage( xfrm, brainMask, template ), 0.5, 1.0, 1, 0 )

      if( doDataAugmentation == TRUE )
        {
        dataAugmentation <- 
          randomlyTransformImageData( template, 
          list( list( warpedT1, warpedT2 ) ),
          list( warpedBrainMask ),
          numberOfSimulations = 1, 
          transformType = 'affineAndDeformation', 
          sdAffine = 0.1,
          deformationTransformType = "bspline",
          numberOfRandomPoints = 1000,
          sdNoise = 2.0,
          numberOfFittingLevels = 4,
          meshSize = 1,
          sdSmoothing = 4.0,
          inputImageInterpolator = 'linear',
          segmentationImageInterpolator = 'nearestNeighbor' )

        simulatedBrainMask <- thresholdImage( dataAugmentation$simulatedSegmentationImages[[1]], 0.5, 1.0, 1, 0 )
        simulatedT1 <- dataAugmentation$simulatedImages[[1]][[1]]
        simulatedT2 <- dataAugmentation$simulatedImages[[1]][[2]]
        simulatedT1Array <- as.array( simulatedT1 )
        simulatedT1Array <- ( simulatedT1Array - mean( simulatedT1Array ) ) / sd( simulatedT1Array )
        simulatedT2Array <- as.array( simulatedT2 )
        simulatedT2Array <- ( simulatedT2Array - mean( simulatedT2Array ) ) / sd( simulatedT2Array )

        # antsImageWrite( simulatedBrainMask, paste0( "./TempData/simBrainMask", i, ".nii.gz" ) )

        X[i,,,,1] <- simulatedT1Array
        X[i,,,,2] <- simulatedT2Array
        Y[i,,,] <- as.integer( as.array( simulatedBrainMask ) )

        i <- i + 1  
        } else {
        warpedT1Array <- as.array( warpedT1 )
        warpedT1Array <- ( warpedT1Array - mean( warpedT1Array ) ) / sd( warpedT1Array )        
        warpedT2Array <- as.array( warpedT2 )
        warpedT2Array <- ( warpedT2Array - mean( warpedT2Array ) ) / sd( warpedT2Array )        

        # antsImageWrite( warpedBrainMask, paste0( "./TempData/wBrainMask", i, ".nii.gz" ) )

        X[i,,,,1] <- warpedT1Array
        X[i,,,,2] <- warpedT2Array
        Y[i,,,] <- as.integer( as.array( warpedBrainMask ) )

        i <- i + 1  
        }
      }
    # stop( "Done testing.")  
    # cat( "\n" )

    encodedY <- encodeUnet( Y, c( 0, 1 ) )

    return( list( X, encodedY ) )
    }
}

File: HemisphereLabeling/Scripts/batch_generator.py
import numpy as np
import random
import ants
import antspynet
import time

def batch_generator(batch_size,
                    template,
                    training_image_files,
                    training_segmentation_files,
                    do_histogram_intensity_warping=True,
                    do_histogram_equalization=True,
                    do_histogram_rank_intensity=True,
                    do_simulate_bias_field=True,
                    do_add_noise=True,
                    do_random_transformation=True,
                    do_random_contralateral_flips=True,
                    do_resampling=True,
                    verbose=False
                    ):

    X = np.zeros((batch_size, *template.shape, 1))
    Y  = np.zeros((batch_size, *template.shape))

    while True:

        batch_count = 0

        while batch_count < batch_size:

            t_idx = random.sample(list(range(len(training_image_files))), 1)[0]

            if verbose:
                print("Batch count: ", batch_count)

            image = ants.image_read(training_image_files[t_idx])
            segmentation = ants.image_read(training_segmentation_files[t_idx])

            center_of_mass_template = ants.get_center_of_mass(template * 0 + 1)
            center_of_mass_image = ants.get_center_of_mass(image * 0 + 1)
            translation = np.round(np.asarray(center_of_mass_image) - np.asarray(center_of_mass_template))
            xfrm = ants.create_ants_transform(transform_type="Euler3DTransform",
                center=np.round(np.asarray(center_of_mass_template)), translation=translation)            
            image = ants.apply_ants_transform_to_image(xfrm, image=image, reference=template, interpolation="linear")
            segmentation = ants.apply_ants_transform_to_image(xfrm, image=segmentation, reference=template, interpolation="nearestNeighbor")

            image = (image - image.min()) / (image.max() - image.min())

            if do_random_transformation:
                tic = time.perf_counter()
                data_augmentation = antspynet.randomly_transform_image_data(template,
                    [[image]],
                    [segmentation],
                    number_of_simulations=1,
                    transform_type='affineAndDeformation',
                    sd_affine=0.075,
                    deformation_transform_type="bspline",
                    number_of_random_points=1000,
                    sd_noise=10.0,
                    number_of_fitting_levels=4,
                    mesh_size=1,
                    sd_smoothing=4.0,
                    input_image_interpolator='linear',
                    segmentation_image_interpolator='nearestNeighbor')
                image = data_augmentation['simulated_images'][0][0]
                segmentation = data_augmentation['simulated_segmentation_images'][0]
                toc = time.perf_counter()
                if verbose:
                    print(f"    Do random transformation {toc - tic:0.4f} seconds.")

            if do_resampling:
                resample_spacing = np.random.uniform(low=1.0, high=1.5, size=3) 
                downsampled_image = ants.resample_image(image, resample_params=resample_spacing,
                                                        use_voxels=False, interp_type=0)
                image = ants.resample_image_to_target(downsampled_image, image)
 
            if do_random_contralateral_flips and random.sample((True, False), 1)[0]:
                tic = time.perf_counter()
                image_array = image.numpy()
                image_array = np.flip(image_array, axis=0)
                image = ants.from_numpy_like(image_array, image)
                segmentation[segmentation == 1] = 3
                segmentation[segmentation == 2] = 1
                segmentation[segmentation == 3] = 2
                segmentation_array = segmentation.numpy()
                segmentation_array = np.flip(segmentation_array, axis=0)
                segmentation = ants.from_numpy_like(segmentation_array, segmentation)
                toc = time.perf_counter()
                if verbose:
                    print(f"    Random contralateral flip {toc - tic:0.4f} seconds")

            if do_histogram_intensity_warping and random.sample((True, True, False), 1)[0]:
                tic = time.perf_counter()
                break_points = [0.2, 0.4, 0.6, 0.8]
                displacements = list()
                for b in range(len(break_points)):
                    displacements.append(abs(random.gauss(0, 0.10)))
                    if random.sample((True, False), 1)[0]:
                        displacements[b] *= -1
                image = antspynet.histogram_warp_image_intensities(image,
                    break_points=break_points, clamp_end_points=(False, False),
                    displacements=displacements)
                image = image.iMath("Normalize")
                toc = time.perf_counter()
                if verbose:
                    print(f"    Histogram warping {toc - tic:0.4f} seconds")

            if do_histogram_equalization and random.sample((False, True), 1)[0]:
                tic = time.perf_counter()
                image = ants.histogram_equalize_image(image)
                image = image.iMath("Normalize")
                toc = time.perf_counter()
                if verbose:
                    print(f"    Histogram equalization {toc - tic:0.4f} seconds")

            if do_histogram_rank_intensity and random.sample((False, True), 1)[0]:
                tic = time.perf_counter()
                image = ants.rank_intensity(image, ants.threshold_image(segmentation, 0, 0, 0, 1))
                image = image.iMath("Normalize")
                toc = time.perf_counter()
                if verbose:
                    print(f"    Histogram rank intensity {toc - tic:0.4f} seconds")

            if do_simulate_bias_field and random.sample((True, True, False), 1)[0]:
                tic = time.perf_counter()
                log_field = antspynet.simulate_bias_field(image, 
                                                            number_of_points=100, 
                                                            sd_bias_field=0.1, 
                                                            number_of_fitting_levels=2, 
                                                            mesh_size=10)
                log_field = log_field.iMath("Normalize")
                field_array = np.power(np.exp(log_field.numpy()), random.sample((2, 3), 1)[0])
                image = image * ants.from_numpy_like(field_array, image)
                image = image.iMath("Normalize")
                toc = time.perf_counter()
                if verbose:
                    print(f"    Simulate bias field {toc - tic:0.4f} seconds")

            if do_add_noise and random.sample((True, True, False), 1)[0]:
                tic = time.perf_counter()
                image_ones = image * 0 + 1                
                noise_parameters = (0.1, random.uniform(0.05, 0.15))
                noise = ants.add_noise_to_image(image_ones, noise_model="additivegaussian", noise_parameters=noise_parameters)
                noise = ants.smooth_image(noise, sigma=random.uniform(0.25, 0.5))
                image = image * noise
                image = image.iMath("Normalize")
                toc = time.perf_counter()
                if verbose:
                    print(f"    Add noise {toc - tic:0.4f} seconds")

            image = ants.iMath_normalize(image) 

            tic = time.perf_counter()

            X[batch_count,:,:,:,0] = image.numpy()
            Y[batch_count,:,:,:] = segmentation.numpy()
                
            batch_count = batch_count + 1    
            if batch_count >= batch_size:
                break

        tic = time.perf_counter()
        encoded_Y = antspynet.encode_unet(Y, np.unique(Y).astype('int32'))
        toc = time.perf_counter()
        if verbose:
            print(f"Encode Y {toc - tic:0.4f} seconds.")

        if verbose:
            yield X, Y
        else:
            yield X, encoded_Y

